### PR TITLE
Feature/PPM-345 create auto configuration task

### PR DESCRIPTION
**Pull request by morantr**
Create an AP Autoconfiguration task.

Passed certification tests:
https://gitlab.com/prpl-foundation/prplMesh/-/pipelines/176499531

### CODE REVIEW

File: agent/src/beerocks/slave/tasks/task.h [hunk 1]
```diff
@@ -48,8 +48,9 @@ class Task {
      * @brief Handle events, which are being send from containing thread.
      * 
      * @param event_enum_value Event enum value which shall be defined on the task.
+     * @param event_obj Pointer to some chunk of memory used to pass data to the event handler.
      */
-    virtual void handle_event(uint8_t event_enum_value){};
+    virtual void handle_event(uint8_t event_enum_value, void *event_obj){};
```
Comment (kantera800) [inline agent/src/beerocks/slave/tasks/task.h]: lets say you have 2 ap_autoconfiguration tasks in the unified agent running in parallel , 
how do you make sure they don't write to the same "chunk of memory"
why not use AgentDB?

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 2]
```diff
@@ -1725,7 +1725,9 @@ bool backhaul_manager::handle_slave_backhaul_message(std::shared_ptr<sRadioInfo>
 
         // If we're already connected, send a notification to the slave
         if (FSM_IS_IN_STATE(OPERATIONAL)) {
-            send_autoconfig_search_message(radio->front.iface_name);
+            m_task_pool.send_event(eTaskType::AP_AUTOCONFIGURATION,
+                                   ApAutoConfigurationTask::eEvent::START_AP_AUTOCONFIGURATION,
+                                   &radio->front.iface_name);
```
Comment (kantera800) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:1701]: you should consider at least pass event data as a const .

File: agent/src/beerocks/slave/tasks/task.h (hunk 1 above)
Comment (morantr) [reply agent/src/beerocks/slave/tasks/task.h]: 1. There is no such thing task of the same type running in parallel on the Agent.
2. Even if it was allowed to run several instances of the same task, they would not run in parallel. They would run one after another. This is how it designed.
3. The chunk of memory is a way to pass some type of object without explicitly declare it on the handle_event signature,
the only difference is that it passed as void pointer instead of explicit type.
Either way, the incoming chuck of memory is for reading, and not writing.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 2 above)
Comment (morantr) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:1701]: Ok, it is no harm to do it.

File: agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp [hunk 3]
```diff
@@ -68,27 +68,79 @@ void ApAutoConfigurationTask::work()
...
+            // the configuration flow. When the son_slave will be unified, the configuration flow
+            // will move to here.
+            FSM_MOVE_STATE(radio_iface, eState::CONFIGIRED);
             break;
         }
         case eState::WAIT_FOR_AP_CONFIGURATION_COMPLETE: {
             break;
         }
         case eState::CONFIGIRED: {
+            ap_autoconfig_not_completed &= false;
```
Comment (kantera800) [inline agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp]:
```suggestion
            ap_autoconfig_not_completed = false;
```

File: tests/environment.py [hunk 4]
```diff
@@ -65,7 +65,8 @@ def prplmesh_command(self, command: str, *args: str) -> bytes:
         '''
```
Comment (kantera800) [inline tests/environment.py:65]: This commit is not related to this PR at all merge it as a separate fixup .
Also fix commit message typos.
`Commit 603b38e adds an option not to fail on log missmatch.`

File: agent/src/beerocks/slave/tasks/task.h (hunk 1 above)
Comment (kantera800) [reply agent/src/beerocks/slave/tasks/task.h]: > They would run one after another. 

that's exactly what i mean and now you are allowing them to use shared resource without protecting it.
>    Either way, the incoming chuck of memory is for reading, and not writing.

the user doesn't know its not for writing. you can't make everybody read instructions all the time.

If you must pass it as a const.

Comment (morantr) [reply agent/src/beerocks/slave/tasks/task.h]: This is not a shared resource, only one task will get it.

File: agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp (hunk 3 above)
Comment (morantr) [reply agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp]: There is a reason I added the `&`, so if one of the radios did not finish the discovery flow, then the autoconfiguration status will not be marked as completed on the database.

File: tests/environment.py (hunk 4 above)
Comment (morantr) [reply tests/environment.py:65]: I need this commit to pass testflows on CI. Without it, it marks it as error on the test log.
fix commit message.

File: agent/src/beerocks/slave/tasks/task.h (hunk 1 above)
Comment (kantera800) [reply agent/src/beerocks/slave/tasks/task.h]: Not true since you are adding it to the task class. 
theoretically both task could be triggered with a pointer to the same "chunk of memory"

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 5]
```diff
@@ -192,6 +194,12 @@ backhaul_manager::backhaul_manager(const config_file::sConfigSlave &config,
         LOG(ERROR) << "failed to allocate Topology Task!";
     }
     m_task_pool.add_task(topology_task);
+
+    auto ap_auto_configuration_task = std::make_shared<ApAutoConfigurationTask>(*this, cmdu_tx);
+    if (!ap_auto_configuration_task) {
+        LOG(ERROR) << "failed to allocate ap_auto_configuration_task Task!";
+    }
+    m_task_pool.add_task(ap_auto_configuration_task);
```
Comment (vitalii-komisarenko) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Probably the task should not be added into the task pool if it failed to allocate.

File: agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+static constexpr uint8_t MAX_FAILED_AUTOCONFIG_SEARCH_ATTEMPTS = 20;
+static constexpr uint8_t AUTOCONFIG_DISCOVERY_TIMEOUT_SECONDS  = 1;
+
+#define FSM_MOVE_STATE(radio_iface, new_state)                                                     \
+    ({                                                                                             \
+        LOG(TRACE) << "AP_AUTOCONFIGURATION " << radio_iface                                       \
+                   << " FSM: " << fsm_state_to_string(m_state[radio_iface].state) << " --> "       \
+                   << fsm_state_to_string(new_state);                                              \
+        m_state[radio_iface].state = new_state;                                                    \
+    })
```
Comment (vitalii-komisarenko) [inline agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp:39]: This kind of macro is used in multiple places. What about moving it to a common header file?

File: agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    case eState::WAIT_FOR_AP_CONFIGURATION_COMPLETE:
+        return "WAIT_FOR_AP_CONFIGURATION_COMPLETE";
+    case eState::CONFIGIRED:
+        return "CONFIGIRED";
+    default:
+        LOG(ERROR) << "state argument doesn't have an enum";
+        break;
+    }
+    return std::string();
+}
```
Comment (vitalii-komisarenko) [inline agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp:61]: It backhaul manager string representation is generated by `const char *backhaul_manager::s_arrStates[] = {FOREACH_STATE(GENERATE_STRING)};`.
Let's unify the code :)

**Review by vitalii-komisarenko — COMMENTED** [review]
Minor comments:
1) in commits 60dcc170e624, 1e3d28b72f4e "to Agent DB" is written twice
2) backhaul manager thread has several useful macros to add some syntax sugar for the state machine. They can be shared in a file called "state_machine_utils.h". A separate task can be created for it, since it is not urgent.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 8]
```diff
@@ -1019,7 +1018,7 @@ bool backhaul_manager::backhaul_fsm_main(bool &skip_select)
 
         db->backhaul.connection_type = AgentDB::sBackhaul::eConnectionType::Invalid;
 
-        controller_bridge_mac.clear();
+        db->controller_info.bridge_mac = {};
```
Comment (itayx) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]:
```suggestion
        db->controller_info.bridge_mac = {.oct = {0}};
```

**Review by itayx — APPROVED** [review]
Looks good, approving but with a small fix

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 5 above)
Comment (morantr) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: OOPS!
Fixing.
I moved the generating of the task to the init function which can fail.

File: agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp (hunk 6 above)
Comment (morantr) [reply agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp:39]: It is similar but not the same. This one has unique implementation that fit only to this task.
Anyway, you can suggest opening a task on creating a common FSM infrastructure. I just don't want to do it as part of this PR.

File: agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp (hunk 7 above)
Comment (morantr) [reply agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp:61]: TBH I hate the whole "FOREACH_STATE(GENERATE_STRING)}" macro on the backhaul manager. We suppose to write code in CPP and not in MACRO.
Using macro should take place only if it has a good purpose.
The only reason I have chosen to use "FSM_MOVE_STATE" is because I wanted that on a state switch, it will print to the log file the line that the switch has occurred. If it was a regular function it would print the same line everywhere.

I'll might unify the code by removing this macro crap on the backhaul manager as part of the unified agent epic. 🙂

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 8 above)
Comment (morantr) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Replaced with ` db->controller_info.bridge_mac = network_utils::ZERO_MAC`

File: agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp (hunk 3 above)
Comment (kantera800) [reply agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp]: if you want indication that both radios finished discovery you should use  `db _statuses.ap_autoconfiguration_completed ` in line 155.
please avoid  `&=` and use something more code readable.

Also, i would add this (or similar) comment to the code as well
`if one of the radios did not finish the discovery flow, then the autoconfiguration status will not be marked as completed on the database`

**Comment by kantera800** [comment]
please make sure you test autoconfiguration on real platform and pass at least 4.8.1 agent certification test.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 9]
```diff
@@ -2825,8 +2828,9 @@ bool backhaul_manager::handle_1905_autoconfiguration_response(ieee1905_1::CmduMe
 {
     LOG(DEBUG) << "received autoconfiguration response message";
     auto db = AgentDB::get();
-    if (!controller_bridge_mac.empty() && src_mac != controller_bridge_mac) {
-        LOG(INFO) << "current controller_bridge_mac=" << controller_bridge_mac
+    if (db->controller_info.bridge_mac != network_utils::ZERO_MAC &&
```
Comment (vitalii-komisarenko) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: I just wonder if it is worth adding automatic conversion of sMacAddr to bool.
Comparing mac addresses to ZERO_MAC often happens often in the code.

File: agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp (hunk 3 above)
Comment (vitalii-komisarenko) [reply agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp]: Did you mean `ap_autoconfig_not_completed  = ap_autoconfig_not_completed  && false` ?
AFAIK only `&&`, `||` and `?` perform lazy evaluation in C++.

**Review by vitalii-komisarenko — CHANGES_REQUESTED** [review]
I agree with @kantera800 that using `&=` with booleans is unclear.
Personally I don't understand the difference between `x = false` and `x &= false`. I am also not sure that bitwise operation on booleans are guaranteed to work as expected in C++.

All other code looks good to me.

File: agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp (hunk 3 above)
Comment (morantr) [reply agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp]: I removed it completely, and replace it with a counter that is incremented on the CONFIGURED state which I compare to the number of radios. This way I also avoid a reverse logic variable name.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 9 above)
Comment (morantr) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: What is the problem with comparing `sMacAddr` to `ZERO_MAC`?
Comparing `sMacAddr` to `false` is weird IMO.

File: agent/src/beerocks/slave/tasks/task.h (hunk 1 above)
Comment (morantr) [reply agent/src/beerocks/slave/tasks/task.h]: We agreed that it is a valid change in a private conversation, therefore mark as resolve.

**Review by kantera800 — APPROVED** [review]

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 9 above)
Comment (vitalii-komisarenko) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: I mean something like this:
`if (db->controller_info.bridge_mac && (src_mac != db->controller_info.bridge_mac))`

**Review by vitalii-komisarenko — APPROVED** [review]
Looks good to me, approved.
